// @nowarp/misti/dist/internals/lattice/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetMeetSemilattice = exports.SetJoinSemilattice = void 0;
/**
 * Implementation of a join semilattice for sets.
 *
 * @template T The type of elements in the sets.
 */
class SetJoinSemilattice {
    /**
     * Joins two sets by union.
     */
    join(a, b) {
        return new Set([...a, ...b]);
    }
    /**
     * Returns the bottom element: empty set.
     */
    bottom() {
        return new Set();
    }
    /**
     * Subset relation: a ≤ b iff a ⊆ b
     */
    leq(a, b) {
        return [...a].every((x) => b.has(x));
    }
}
exports.SetJoinSemilattice = SetJoinSemilattice;
/**
 * Implementation of a meet semilattice for sets.
 *
 * @template T The type of elements in the sets.
 */
class SetMeetSemilattice {
    /**
     * Meets two sets by intersection.
     */
    meet(a, b) {
        return new Set([...a].filter((x) => b.has(x)));
    }
    /**
     * Returns the top element: empty set.
     */
    top() {
        return new Set();
    }
    /**
     * Reverse subset relation: a ≤ b iff b ⊆ a
     */
    leq(a, b) {
        return [...b].every((x) => a.has(x));
    }
}
exports.SetMeetSemilattice = SetMeetSemilattice;
