// @nowarp/souffle/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SoufflePrettyPrinter = exports.SouffleExecutionError = exports.SouffleInternalError = exports.SouffleUsageError = exports.SouffleError = exports.SouffleEmitter = exports.SouffleContext = exports.binaryConstraint = exports.program = exports.rule = exports.body = exports.booleanConstraint = exports.containsConstraint = exports.matchConstraint = exports.atom = exports.relation = exports.fact = exports.comment = void 0;
__exportStar(require("./executor/index"), exports);
var syntaxConstructors_1 = require("./syntaxConstructors");
Object.defineProperty(exports, "comment", { enumerable: true, get: function () { return syntaxConstructors_1.comment; } });
Object.defineProperty(exports, "fact", { enumerable: true, get: function () { return syntaxConstructors_1.fact; } });
Object.defineProperty(exports, "relation", { enumerable: true, get: function () { return syntaxConstructors_1.relation; } });
Object.defineProperty(exports, "atom", { enumerable: true, get: function () { return syntaxConstructors_1.atom; } });
Object.defineProperty(exports, "matchConstraint", { enumerable: true, get: function () { return syntaxConstructors_1.matchConstraint; } });
Object.defineProperty(exports, "containsConstraint", { enumerable: true, get: function () { return syntaxConstructors_1.containsConstraint; } });
Object.defineProperty(exports, "booleanConstraint", { enumerable: true, get: function () { return syntaxConstructors_1.booleanConstraint; } });
Object.defineProperty(exports, "body", { enumerable: true, get: function () { return syntaxConstructors_1.body; } });
Object.defineProperty(exports, "rule", { enumerable: true, get: function () { return syntaxConstructors_1.rule; } });
Object.defineProperty(exports, "program", { enumerable: true, get: function () { return syntaxConstructors_1.program; } });
Object.defineProperty(exports, "binaryConstraint", { enumerable: true, get: function () { return syntaxConstructors_1.binaryConstraint; } });
var context_1 = require("./context");
Object.defineProperty(exports, "SouffleContext", { enumerable: true, get: function () { return context_1.SouffleContext; } });
var emitter_1 = require("./emitter");
Object.defineProperty(exports, "SouffleEmitter", { enumerable: true, get: function () { return emitter_1.SouffleEmitter; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "SouffleError", { enumerable: true, get: function () { return errors_1.SouffleError; } });
Object.defineProperty(exports, "SouffleUsageError", { enumerable: true, get: function () { return errors_1.SouffleUsageError; } });
Object.defineProperty(exports, "SouffleInternalError", { enumerable: true, get: function () { return errors_1.SouffleInternalError; } });
Object.defineProperty(exports, "SouffleExecutionError", { enumerable: true, get: function () { return errors_1.SouffleExecutionError; } });
var prettyPrinter_1 = require("./prettyPrinter");
Object.defineProperty(exports, "SoufflePrettyPrinter", { enumerable: true, get: function () { return prettyPrinter_1.SoufflePrettyPrinter; } });
